### PR TITLE
added docker check to workflow

### LINKED ISSUE #2402: GitHub PR Workflow Check: Talawa Admin app starts `(Docker)`
**Describe the bug**

The GitHub PR workflow does not check whether the Talawa Admin app can start in Docker

**To Reproduce**

1. Make a change to the repo that makes the Talawa Admin app fail to start in Docker
1. Trigger the workflow to run as part of a PR
1. The workflow **_passes_**

and as expected:

1. Make a change to the repo that makes the Talawa Admin app start in Docker
1. Trigger the workflow to run as part of a PR
1. The workflow passes

**Expected behavior**

1. Make a change to the repo that makes the Talawa Admin app fail to start in Docker
1. Trigger the workflow to run as part of a PR
1. The workflow **_fails_**

and as expected:

1. Make a change to the repo that makes the Talawa Admin app start in Docker
1. Trigger the workflow to run as part of a PR
1. The workflow passes

We now have a test data that can be loaded that resides in the repo. This should be used as part of the test.

**Actual behavior**

- The Talawa Admin app isn't started and tested in the PR workflow

**Screenshots**

- N/A

**Additional details**

1. It must verify starting with Docker
1. Use this issue for reference:
    1. https://github.com/PalisadoesFoundation/talawa-api/issues/1343

**Potential internship candidates**

- Please read this if you are planning to apply for a Palisadoes Foundation internship https://github.com/PalisadoesFoundation/talawa/issues/359

### PR DESCRIPTION
**Pull request by VanshikaSabharwal**
**What kind of change does this PR introduce?**

 bugfix, feature request

**Issue Number:**

Fixes #2402 

**Did you add tests for your changes?**

No

**Snapshots/Videos:**

image

**If relevant, did you update the documentation?**

Not relevent

**Summary**

**Does this PR introduce a breaking change?**

No

**Other information**

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-admin/blob/master/CONTRIBUTING.md)?**

Yes

## Release Notes

- **New Features**
  - Introduced a new job in the GitHub Actions workflow to verify the Talawa Admin application starts correctly in Docker.
  - Added a `Dockerfile` for streamlined application deployment.
  - Included a new section in the installation guide detailing Docker setup, covering prerequisites, development, and production configurations.

- **Enhancements**
  - Updated `.dockerignore` to optimize Docker image builds by excluding unnecessary files.
  - Modified server configuration to change the default port to 4321 and prevent automatic browser opening.

These changes aim to improve the development and deployment experience for users.

### CODE REVIEW

File: .github/workflows/pull-request.yml [hunk 1]
```diff
@@ -253,6 +253,35 @@ jobs:
...
+        
+             - name: Set up Docker
+             run: |
+                  sudo apt-get update
+                  sudo apt-get install -y docker.io
+        
+              - name: Run Docker Container
+                run: |
+                  docker run -d --name talawa-admin-container -p 8080:8080 talawa-admin-app
+
```
[the remark this answers is not in the document]
Comment (VanshikaSabharwal) [reply .github/workflows/pull-request.yml:271]: i did it. 
image

File: .github/workflows/pull-request.yml [hunk 2]
```diff
@@ -253,6 +253,35 @@
...
+          runs-on: ubuntu-latest
+          steps:
+             - name: Checkout the Repository
+             uses: actions/checkout@v4
+        
+             - name: Set up Docker
+             run: |
+                  sudo apt-get update
+                  sudo apt-get install -y docker.io
+
```
[the remark this answers is not in the document]
Comment (VanshikaSabharwal) [reply .github/workflows/pull-request.yml:267]: did it.

File: .github/workflows/pull-request.yml [hunk 3]
```diff
@@ -253,6 +253,35 @@
...
+              - name: Run Docker Container
+                run: |
+                  docker run -d --name talawa-admin-container -p 8080:8080 talawa-admin-app
+        
+              - name: Check if Talawa Admin App is running
+               run: |
+                nc -z localhost 8080
+        
+                curl --fail --silent http://localhost:8080/index.html || exit 1
+
```
[the remark this answers is not in the document]
Comment (VanshikaSabharwal) [reply .github/workflows/pull-request.yml:277]: did  it.

File: .github/workflows/pull-request.yml [hunk 4]
```diff
@@ -253,6 +253,54 @@ jobs:
...
+  Docker-Start-Check:
+          name: Check if Talawa Admin app starts in Docker
+          runs-on: ubuntu-latest
+          needs: [Code-Quality-Checks]
+          steps:
+             - name: Checkout the Repository
+             uses: actions/checkout@v4
+        
+             - name: Set up Docker
+             uses: docker/setup-buildx-action@v3
```
[the remark this answers is not in the document]
Comment (VanshikaSabharwal) [reply .github/workflows/pull-request.yml:265]: i did it. 
image

File: .github/workflows/pull-request.yml [hunk 5]
```diff
@@ -253,6 +253,39 @@
...
+            - name: Run Docker Container
+                run: |
+                  docker run -d --name talawa-admin-app-container -p 8080:8080 talawa-admin-app
+
+            - name: Check if Talawa Admin app starts in Docker
+                runs-on: ubuntu-latest
+                needs: [Code-Quality-Checks]
+                steps:
+                  - name: Checkout the Repository
+                    uses: actions/checkout@v4
```
[the remark this answers is not in the document]
Comment (VanshikaSabharwal) [reply .github/workflows/pull-request.yml:280]: before i did that only and you told me to do this. you are confusing me. https://github.com/PalisadoesFoundation/talawa-admin/pull/2414

[the remark this answers is not in the document]
Comment (VanshikaSabharwal) [reply .github/workflows/pull-request.yml:280]: yes i did it

File: .github/workflows/pull-request.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (VanshikaSabharwal) [reply .github/workflows/pull-request.yml]: i did it

**Review by varshith257 — CHANGES_REQUESTED** [review]
It seems workflow added isn't triggered?

**Review by pranshugupta54 — CHANGES_REQUESTED** [review]
Looks like there's an error in the YAML file—possibly due to formatting or indentation.

Check: https://github.com/PalisadoesFoundation/talawa-admin/actions/runs/11741762559/workflow#L260

**Comment by VanshikaSabharwal** [comment]
> Looks like there's an error in the YAML file—possibly due to formatting or indentation.
> 
> Check: https://github.com/PalisadoesFoundation/talawa-admin/actions/runs/11741762559/workflow#L260

Yes you are right, there is indentation error due to which the workflow is not triggered. I am working on it.

**Comment by varshith257** [comment]
@palisadoes I am still surprised as we don't have any Dockerfile in talawa-admin and how we can support this workflow though 😄

**Comment by VanshikaSabharwal** [comment]
> @palisadoes I am still surprised as we don't have any Dockerfile in talawa-admin and how we can support this workflow though 😄

Hey @varshith257, Do i have to create a DockerFile ?

**Comment by varshith257** [comment]
Check #1075

**Comment by VanshikaSabharwal** [comment]
> Check #1075

Yes, @varshith257 I checked this and i found out that there still isn't any file. so at this point can i make a docker file?

**Comment by palisadoes** [comment]
@VanshikaSabharwal

You'll need to update this PR with:

1. The Dockerfile for it to achieve our goals
2. Changes to INSTALLATION.md so that people know what to do.

There is another issue to test whether the app starts. You'll need to figure out a way for the app starting test and docker app starting test can work together.

**Comment by palisadoes** [comment]
During the week of November 11, 2024 we will start a code freeze on the `develop` branches in Talawa, Talawa Admin and Talawa-API. 

We have completed a project to convert the Talawa-API backend to use PostgreSQL. Work will then begin with us merging code in the `develop` branches to a new `develop-postrgres` branch in each repository.

Planning activities for this will be managed in our #talawa-projects slack channel. A GitHub project will be created to track specially labeled issues. We completed a similar exercise last year using a similar methodology. 

Starting November 12, California time no new PRs will be accepted against the `develop` branch. They must be applied to the `develop-postrgres` branch. 

There are some GSoC project features that will need to be merged into develop. These will be the only exceptions.

This activity and the post GSoC 2024 start date was announced in our #general Slack channel last month as a pinned post.

**Comment by VanshikaSabharwal** [comment]
> @VanshikaSabharwal
> 
> You'll need to update this PR with:
> 
> 1. The Dockerfile for it to achieve our goals
> 2. Changes to INSTALLATION.md so that people know what to do.
> 
> There is another issue to test whether the app starts. You'll need to figure out a way for the app starting test and docker app starting test can work together.

Yes, @palisadoes 

Thank you for the feedback!

1. I'll add the Dockerfile to ensure we meet our goals.
2. I'll update INSTALLATION.md with the necessary steps for Docker setup so users can follow along easily.
3. For the app startup tests, I'll work on a solution that allows both the app starting test and the Docker app starting test to function together without conflicts.

Please let me know if there are any additional requirements. Thanks!

**Comment by VanshikaSabharwal** [comment]
> During the week of November 11, 2024 we will start a code freeze on the `develop` branches in Talawa, Talawa Admin and Talawa-API.
> 
> We have completed a project to convert the Talawa-API backend to use PostgreSQL. Work will then begin with us merging code in the `develop` branches to a new `develop-postrgres` branch in each repository.
> 
> Planning activities for this will be managed in our #talawa-projects slack channel. A GitHub project will be created to track specially labeled issues. We completed a similar exercise last year using a similar methodology.
> 
> Starting November 12, California time no new PRs will be accepted against the `develop` branch. They must be applied to the `develop-postrgres` branch.
> 
> There are some GSoC project features that will need to be merged into develop. These will be the only exceptions.
> 
> This activity and the post GSoC 2024 start date was announced in our #general Slack channel last month as a pinned post.

Thank you for the update!

Understood regarding the code freeze and the branch updates. I’ll make sure to redirect any new PRs to the develop-postgres branch starting November 12. For any GSoC-related features needing to go into develop, I’ll follow up as needed to confirm the specifics.

I’ll also keep an eye on #talawa-projects in Slack and the GitHub project board to stay aligned with the planning and tracking.

File: docker-compose.yaml [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+  app:
+    build:
+      context: .
+      dockerfile: Dockerfile
+    environment:
+      - REACT_APP_TALAWA_URL=${REACT_APP_TALAWA_URL}
+    # volumes:
+    # - .:/usr/src/app
+    ports:
+      - 4321:4321
```
[the remark this answers is not in the document]
Comment (VanshikaSabharwal) [reply docker-compose.yaml:11]: Did it. 
image

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,15 @@
+FROM node:20.10.0 AS build
+
+WORKDIR /usr/src/app
+
+COPY . .
+
```
[the remark this answers is not in the document]
Comment (VanshikaSabharwal) [reply Dockerfile:6]: did it.

File: Dockerfile [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+RUN npm install -g typescript
+
+RUN npm install
+
```
[the remark this answers is not in the document]
Comment (VanshikaSabharwal) [reply Dockerfile:10]: Did it. 
image

File: docker-compose.yaml [hunk 9]
```diff
@@ -0,0 +1,21 @@
+services:
+  app:
+    build:
+      context: .
+      dockerfile: Dockerfile
+    environment:
+      - REACT_APP_TALAWA_URL=${REACT_APP_TALAWA_URL}
+    ports:
+      - 4321:4321
```
[the remark this answers is not in the document]
Comment (VanshikaSabharwal) [reply docker-compose.yaml:9]: did it. 
image

**Comment by VanshikaSabharwal** [comment]
I have merged https://github.com/PalisadoesFoundation/talawa-admin/pull/2418#event-15244530247 into this PR as suggested by @palisadoes.

File: docker-compose.yaml [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+      - REACT_APP_TALAWA_URL=${REACT_APP_TALAWA_URL:?Error: REACT_APP_TALAWA_URL is required}
+    ports:
+      - 4321:4321
+    user: node
+    healthcheck:
+      test: ['CMD', 'curl', '-f', 'http://localhost:4321']
+      interval: 30s
+      timeout: 10s
+      retries: 5
+      start_period: 40s
```
[the remark this answers is not in the document]
Comment (VanshikaSabharwal) [reply docker-compose.yaml:16]: did it. 
image

File: .github/workflows/pull-request.yml [hunk 11]
```diff
@@ -253,6 +253,65 @@ jobs:
...
+              echo "Still waiting for app to start... ${timeout}s remaining"
+            fi
+          done
+          if [ $timeout -eq 0 ]; then
+            echo "Timeout waiting for application to start"
+            exit 1
+          fi
+          echo "Port check passed, verifying health endpoint..."
+          curl --fail --silent http://localhost:4321/health || exit 1
+          echo "Health check passed successfully"
```
[the remark this answers is not in the document]
Comment (VanshikaSabharwal) [reply .github/workflows/pull-request.yml:305]: did it. 
image

**Comment by VanshikaSabharwal** [comment]
Now this PR also solves https://github.com/PalisadoesFoundation/talawa-admin/issues/1075 issue by adding Dockerfile,.docker-ignore and docker-compose file.

**Comment by palisadoes** [comment]
1. Remember the INSTALLATION.md file
2. Why is the docker check failing?

**Comment by VanshikaSabharwal** [comment]
> 1. Remember the INSTALLATION.md file
> 2. Why is the docker check failing?

Yes, I will update INSTALLATION.md and also i think that there is no health check point due to which the docker test is failing.

File: .github/workflows/pull-request.yml [hunk 12]
```diff
@@ -240,19 +240,76 @@
...
+              echo "Still waiting for app to start... ${timeout}s remaining"
+            fi
+          done
+
+          if [ $timeout -eq 0 ]; then
+            echo "Timeout waiting for application to start"
+            echo "Container logs:"
+            docker logs talawa-admin-app-container
+            exit 1
+          fi
```
[the remark this answers is not in the document]
Comment (VanshikaSabharwal) [reply .github/workflows/pull-request.yml:304]: there is no health check point in app

File: .github/workflows/pull-request.yml [hunk 13]
```diff
@@ -240,19 +240,78 @@ jobs:
...
+            echo "Still waiting for app to start... ${timeout}s remaining"
+          fi
+        done
+
+        if [ $timeout -eq 0 ]; then
+          echo "Timeout waiting for application to start"
+          echo "Container logs:"
+          docker logs talawa-admin-app-container
+          exit 1
+        fi
```
[the remark this answers is not in the document]
Comment (VanshikaSabharwal) [reply .github/workflows/pull-request.yml]: did it

**Comment by VanshikaSabharwal** [comment]
@palisadoes I have updated the INSTALLATION.md file and deleted the docker-compose.yml file, as I thought there was no specific reason to keep it, and it was causing issues.

File: INSTALLATION.md [hunk 14]
```diff
@@ -145,6 +148,28 @@ npm install
...
+## Run the Docker container:
+
+After the build is complete, run the Docker container using this command:
+
+```
+docker run -p 4321:4321 talawa-admin
+```
+
+This will expose the container's port 4321 on your local machine, making the application accessible at http://localhost:4321.
+
```
[the remark this answers is not in the document]
Comment (VanshikaSabharwal) [reply INSTALLATION.md]: did it

File: .github/workflows/pull-request.yml [hunk 15]
```diff
@@ -240,19 +240,76 @@ jobs:
...
-          
+
           # Clone the specified repository using the extracted branch name
-          git clone --branch $FULL_BRANCH_NAME https://github.com/PalisadoesFoundation/talawa-api && ls -a 
+          git clone --branch $FULL_BRANCH_NAME https://github.com/PalisadoesFoundation/talawa-api && ls -a
 
       - name: Validate Documents
         run: graphql-inspector validate './src/GraphQl/**/*.ts' './talawa-api/schema.graphql'
 
+  Docker-Start-Check:
```
Comment (palisadoes) [inline .github/workflows/pull-request.yml:256]: Please ensure that this does not run if the PR is submitted by the dependabot user

**Review by palisadoes — CHANGES_REQUESTED** [review]
Please see requested changes

File: .github/workflows/pull-request.yml (hunk 15 above)
Comment (VanshikaSabharwal) [reply .github/workflows/pull-request.yml:256]: Yes, I did it. 
image

File: .github/workflows/pull-request.yml [hunk 16]
```diff
@@ -240,19 +240,77 @@ jobs:
...
+        uses: actions/checkout@v4
+
+      - name: Set up Docker
+        uses: docker/setup-buildx-action@v3
+        with:
+          driver-opts: |
+            image=moby/buildkit:latest
+          cache-from: type=gha
+          cache-to: type=gha,mode=max
+
```
[the remark this answers is not in the document]
Comment (VanshikaSabharwal) [reply .github/workflows/pull-request.yml:272]: I removed as suggested. 
image

File: .github/workflows/pull-request.yml [hunk 17]
```diff
@@ -240,19 +240,77 @@
...
+          done
+
+          if [ $timeout -eq 0 ]; then
+            echo "Timeout waiting for application to start"
+            echo "Container logs:"
+            docker logs talawa-admin-app-container
+            exit 1
+          fi
+          echo "Port check passed, verifying health endpoint..."
+
```
[the remark this answers is not in the document]
Comment (VanshikaSabharwal) [reply .github/workflows/pull-request.yml:307]: +         echo "Port check passed, verifying application..."
+         # Wait for the application to fully initialize
+         sleep 5
+         # Verify the application is responding
+         if ! curl --fail --silent --max-time 10 "http://localhost:4321/"; then
+           echo "Application health check failed"
+           docker logs talawa-admin-app-container
+           exit 1
+         fi
+         echo "Application is healthy". these lines were causing issue as we do not have a health check point and the workflow was failing even when the talawa-app was running in docker

File: INSTALLATION.md [hunk 18]
```diff
@@ -145,6 +148,38 @@ npm install
...
 The prerequisites are now installed. The next step will be to get the app up and running.
 
+# Installation using Docker
+
+## Prerequisites
+
+1. Install Docker on your system:
+   - [Docker Desktop for Windows/Mac](https://www.docker.com/products/docker-desktop)
+   - [Docker Engine for Linux](https://docs.docker.com/engine/install/)
+
```
[the remark this answers is not in the document]
Comment (VanshikaSabharwal) [reply INSTALLATION.md:158]: I removed the docker compose as it was not needed

File: .github/workflows/pull-request.yml [hunk 19]
```diff
@@ -240,19 +240,75 @@
...
```
[the remark this answers is not in the document]
Comment (VanshikaSabharwal) [reply .github/workflows/pull-request.yml:305]: I did it. 
image

File: INSTALLATION.md [hunk 20]
```diff
@@ -145,24 +129,58 @@ npm install
...
+3.  Run the Docker container:
+
+After the build is complete, run the Docker container using this command:
+
+```bash
+docker run -p 4321:4321 talawa-admin
+```bash
+
+The application will be accessible at `http://localhost:4321`
+
```
[the remark this answers is not in the document]
Comment (VanshikaSabharwal) [reply INSTALLATION.md:163]: did it. 
image

**Comment by VanshikaSabharwal** [comment]
This is video demonstrating that everything is working well
https://github.com/user-attachments/assets/83498abd-e3b7-4dff-b4d8-c61cfda522dd

**Comment by VanshikaSabharwal** [comment]
I have video of building docker image as well but video is too long

**Review by palisadoes — APPROVED** [review]

**Comment by palisadoes** [comment]
@VanshikaSabharwal Please open another PR with the same changes against the `develop-postgres` branch

**Comment by VanshikaSabharwal** [comment]
> @VanshikaSabharwal Please open another PR with the same changes against the `develop-postgres` branch

okay

**Pull request merged**